qubesadmin/tools/qvm_template.py

```diff
@@ -0,0 +1,407 @@
+                package_hdr[rpm.RPMTAG_EPOCHNUM]
+            tpl.features['template-version'] = \
+                package_hdr[rpm.RPMTAG_VERSION]
+            tpl.features['template-release'] = \
+                package_hdr[rpm.RPMTAG_RELEASE]
+            tpl.features['template-install-date'] = \
+                str(datetime.datetime.today())
+            tpl.features['template-name'] = name
+            # TODO: Store source repo
+
+def qrexec_popen(args, app, service, stdout=subprocess.PIPE, filter_esc=True):
+    if args.updatevm:
+        return app.domains[args.updatevm].run_service(
+            service,
+            filter_esc=filter_esc,
+            stdout=stdout)
+    else:
+        return subprocess.Popen([
+                '/etc/qubes-rpc/%s' % service,
+            ],
+            stdin=subprocess.PIPE,
+            stdout=stdout,
+            stderr=subprocess.PIPE)
+
+def qrexec_payload(args, app, spec):
+    payload = ''
+    for r in args.enablerepo if args.enablerepo else []:
+        payload += '--enablerepo=%s\n' % r
+    for r in args.disablerepo if args.disablerepo else []:
+        payload += '--disablerepo=%s\n' % r
+    for r in args.repoid if args.repoid else []:
+        payload += '--repoid=%s\n' % r
+    payload += '--releasever=%s\n' % args.releasever
+    payload += spec + '\n'
+    payload += '---\n'
+    for fn in args.repo_files:
+        with open(fn, 'r') as f:
+            payload += f.read() + '\n'
+    return payload
+
+def qrexec_repoquery(args, app, spec='*'):
+    proc = qrexec_popen(args, app, 'qubes.TemplateSearch')
+    payload = qrexec_payload(args, app, spec)
+    stdout, stderr = proc.communicate(payload.encode('UTF-8'))
+    stdout = stdout.decode('ASCII')
+    if proc.wait() != 0:
+        return None
+    result = []
+    for line in stdout.strip().split('\n'):
+        entry = line.split(':')
+        if not entry[0].startswith(PACKAGE_NAME_PREFIX):
+            continue
+        entry[0] = entry[0][len(PACKAGE_NAME_PREFIX):]
+        result.append(entry)
+    return result
+
+def qrexec_download(args, app, spec, path, dlsize=None):
+    with open(path, 'wb') as f:
+        # Don't filter ESCs for binary files
+        proc = qrexec_popen(args, app, 'qubes.TemplateDownload',
+            stdout=f, filter_esc=False)
+        payload = qrexec_payload(args, app, spec)
+        proc.stdin.write(payload.encode('UTF-8'))
+        proc.stdin.close()
+        while proc.poll() == None:
+            width = shutil.get_terminal_size((80, 20)).columns
+            pct = '%5.f%% ' % math.floor(f.tell() / dlsize * 100)
+            bar_len = width - len(pct) - 2
+            num_hash = math.floor(f.tell() / dlsize * bar_len)
+            num_space = bar_len - num_hash
+            # Clear previous bar
+            print(u'\u001b[1000D', end='', file=sys.stderr)
+            print(pct + '[' + ('#' * num_hash) + (' ' * num_space) + ']',
+                end='', file=sys.stderr)
```

Comment:
Take a look at tqdm python module for progress bar. New (not released yet) pdf converter uses it: https://github.com/QubesOS/qubes-app-linux-pdf-converter/blob/master/qubespdfconverter/client.py

Reply:
Ahh, so tqdm is actually available :) Mentioned it above but was not sure if it's okay to pull in a dependency for this.